README.md
# Chef Cookbook Localmail
[![Build Status](https://travis-ci.org/codenamephp/chef.cookbook.localmail.svg?branch=dev)](https://travis-ci.org/codenamephp/chef.cookbook.localmail)

Cookbook that installs a local mailserver and mail client for local only mail used during development.

## Requirements

### Supported Platforms

- Debian Buster (probably works with older versions as well)

### Chef

- Chef 13.0+

### Cookbook Depdendencies

- apt
- codenamephp_docker

## Usage

Add the cookbook to your Berksfile:

```ruby
cookbook 'codenamephp_localmail'
```


Build a wrapper cookbook and use the resources as needed.

### Resources

#### Mailhog

Mailhog is a mail trap for local development. It offers a nice
web UI and also advanced features like a JSON REST API.

This resource installs mailhog as docker container to avoid any other
build or runtime dependencies (short of docker of course, but chances are
docker is already used).

It also install a custom sendmail script that relays the calls to the
sendmail within the docker container.

##### Properties
- `webui_port` (Integer): The port on the host whre the web ui will be available at, defaults to `8025`
- `sendmail_install_path` (String): The path to where the custom sendmail is installed to, defaults to `/usr/sbin/sendmail`

##### Examples

With minimal properties:
```ruby
# Install
codenamephp_localmail_mailhog 'install mailhog'

# Uninstall
codenamephp_localmail_mailhog 'uninstall mailhog' do
  action :uninstall
end
```
With custom port and path:
```ruby
# Install
codenamephp_localmail_mailhog 'install mailhog' do
  webui_port 1234
  sendmail_install_path '/some/other/path'
end

# Uninstall - webui_port is not relevant here
codenamephp_localmail_mailhog 'uninstall mailhog' do
  sendmail_install_path '/some/ohter/path'
end
```

CHANGELOG.md
# Changelog

## [3.4.1](https://github.com/codenamephp/chef.cookbook.localmail/tree/3.4.1) (2021-12-29)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/3.4.0...3.4.1)

**Merged pull requests:**

- Bump bobheadxi/deployments from 0.6.1 to 0.6.2 [\#28](https://github.com/codenamephp/chef.cookbook.localmail/pull/28) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump tubone24/update\_release from 1.2.0 to 1.3.1 [\#27](https://github.com/codenamephp/chef.cookbook.localmail/pull/27) ([dependabot[bot]](https://github.com/apps/dependabot))
- Bump bobheadxi/deployments from 0.6.0 to 0.6.1 [\#26](https://github.com/codenamephp/chef.cookbook.localmail/pull/26) ([dependabot[bot]](https://github.com/apps/dependabot))

## [3.4.0](https://github.com/codenamephp/chef.cookbook.localmail/tree/3.4.0) (2021-10-09)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/3.3.0...3.4.0)

**Implemented enhancements:**

- Updated docker version to 9.0 [\#24](https://github.com/codenamephp/chef.cookbook.localmail/pull/24) ([bastianschwarz](https://github.com/bastianschwarz))

## [3.3.0](https://github.com/codenamephp/chef.cookbook.localmail/tree/3.3.0) (2021-09-11)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/3.2.0...3.3.0)

**Implemented enhancements:**

- Added unified\_mode [\#21](https://github.com/codenamephp/chef.cookbook.localmail/pull/21) ([bastianschwarz](https://github.com/bastianschwarz))

## [3.2.0](https://github.com/codenamephp/chef.cookbook.localmail/tree/3.2.0) (2020-09-29)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/v3.1.0...3.2.0)

**Implemented enhancements:**

- Updated build [\#18](https://github.com/codenamephp/chef.cookbook.localmail/pull/18) ([bastianschwarz](https://github.com/bastianschwarz))

**Fixed bugs:**

- Update build [\#16](https://github.com/codenamephp/chef.cookbook.localmail/issues/16)

**Merged pull requests:**

- \[CHANGELOG\] 3.2.0 [\#19](https://github.com/codenamephp/chef.cookbook.localmail/pull/19) ([github-actions[bot]](https://github.com/apps/github-actions))
- \[RELEASE 3.2.0\] [\#17](https://github.com/codenamephp/chef.cookbook.localmail/pull/17) ([bastianschwarz](https://github.com/bastianschwarz))

## [v3.1.0](https://github.com/codenamephp/chef.cookbook.localmail/tree/v3.1.0) (2020-03-01)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/v3.0.1...v3.1.0)

**Closed issues:**

- Use sendmail from docker instead of extra mhsendmail [\#14](https://github.com/codenamephp/chef.cookbook.localmail/issues/14)

**Merged pull requests:**

- \[RELEASE 3.1.0\] [\#15](https://github.com/codenamephp/chef.cookbook.localmail/pull/15) ([bastianschwarz](https://github.com/bastianschwarz))

## [v3.0.1](https://github.com/codenamephp/chef.cookbook.localmail/tree/v3.0.1) (2020-03-01)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/v2.0.0...v3.0.1)

**Closed issues:**

- Use mailhog [\#11](https://github.com/codenamephp/chef.cookbook.localmail/issues/11)
- Remove postfix [\#10](https://github.com/codenamephp/chef.cookbook.localmail/issues/10)

**Merged pull requests:**

- \[RELEASE 3.0.0\] [\#12](https://github.com/codenamephp/chef.cookbook.localmail/pull/12) ([bastianschwarz](https://github.com/bastianschwarz))

## [v2.0.0](https://github.com/codenamephp/chef.cookbook.localmail/tree/v2.0.0) (2018-09-13)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/v1.0.1...v2.0.0)

**Implemented enhancements:**

- Switch to postfix again ... [\#3](https://github.com/codenamephp/chef.cookbook.localmail/issues/3)

**Merged pull requests:**

- \[RELEASE 3.0.1\] [\#13](https://github.com/codenamephp/chef.cookbook.localmail/pull/13) ([bastianschwarz](https://github.com/bastianschwarz))
- \[RELEASE 2.0.0\] [\#7](https://github.com/codenamephp/chef.cookbook.localmail/pull/7) ([bastianschwarz](https://github.com/bastianschwarz))
- \[RELEASE 2.0.0\] [\#5](https://github.com/codenamephp/chef.cookbook.localmail/pull/5) ([bastianschwarz](https://github.com/bastianschwarz))

## [v1.0.1](https://github.com/codenamephp/chef.cookbook.localmail/tree/v1.0.1) (2018-09-08)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/v1.0.0...v1.0.1)

**Merged pull requests:**

- \[RELEASE 1.0.1\] [\#2](https://github.com/codenamephp/chef.cookbook.localmail/pull/2) ([bastianschwarz](https://github.com/bastianschwarz))

## [v1.0.0](https://github.com/codenamephp/chef.cookbook.localmail/tree/v1.0.0) (2018-09-08)

[Full Changelog](https://github.com/codenamephp/chef.cookbook.localmail/compare/783fe4617cb3f3c8aad212ba09018b31f042d54e...v1.0.0)

**Merged pull requests:**

- \[RELEASE 1.0.0\] [\#1](https://github.com/codenamephp/chef.cookbook.localmail/pull/1) ([bastianschwarz](https://github.com/bastianschwarz))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


spec/unit/recipes/default_spec.rb
#
# Cookbook:: codenamephp_localmail
# Spec:: default
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'codenamephp_localmail::default' do
  context 'When all attributes are default' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'installs the default jre' do
      expect(chef_run).to install_package('default-jre')
    end

    it 'installs the jar file' do
      expect(chef_run).to create_remote_file('Download MockMock.jar').with(
        source: 'https://github.com/tweakers-dev/MockMock/blob/master/release/MockMock.jar?raw=true',
        path: '/var/opt/MockMock.jar',
        owner: 'root',
        group: 'root',
        mode: 0o755
      )
    end

    it 'installs the init script' do
      expect(chef_run).to create_template('Manage service startup script').with(
        source: 'mockmock.erb',
        path: '/etc/init.d/mockmock',
        owner: 'root',
        group: 'root',
        mode: 0o755
      )
    end

    it 'starts and enables the service' do
      expect(chef_run).to start_service('mockmock')
      expect(chef_run).to enable_service('mockmock')
    end
  end
end


spec/unit/recipes/postfix_spec.rb
#
# Cookbook:: codenamephp_localmail
# Spec:: postfix
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'codenamephp_localmail::postfix' do
  context 'When all attributes are default' do
    let(:chef_run) do
      runner = ChefSpec::SoloRunner.new
      runner.converge(described_recipe)
    end
    let(:install_makefile) { chef_run.cookbook_file('install makefile for managing configurations') }
    let(:install_chef_header) { chef_run.template('apply chef header config') }
    let(:install_local_cf) { chef_run.template('apply local config') }
    let(:install_local_destinations_table) { chef_run.cookbook_file('copy destination table for local config') }
    let(:copy_main_cf) { chef_run.file('copy default main.cf to dropfolder') }
    let(:generate_config) { chef_run.execute('generate config') }

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end

    it 'includes apt cookbook' do
      expect(chef_run).to include_recipe('apt')
    end

    it 'will insall postfix from package with preseed' do
      expect(chef_run).to install_package('postfix').with(response_file: 'postfix/preseed.erb')
    end

    it 'will install postfix-pcre package' do
      expect(chef_run).to install_package('postfix-pcre')
    end

    it 'will install make from package' do
      expect(chef_run).to install_package('make')
    end

    it 'will install the makefile for config management and notify config generetion' do
      expect(chef_run).to create_cookbook_file('install makefile for managing configurations')
      expect(install_makefile).to notify('execute[generate config]').to(:run).delayed
    end

    it 'sets up config generation' do
      expect(chef_run).to nothing_execute('generate config').with(
        command: 'make',
        cwd: '/etc/postfix'
      )
      expect(generate_config).to notify('service[postfix]').to(:restart).delayed
    end

    it 'will create main.cf.d dropfolder' do
      expect(chef_run).to create_directory('create main.cf.d dropfolder').with(
        path: '/etc/postfix/main.cf.d',
        mode: 0o755
      )
    end

    it 'will copy default main.cf to dropfolder and regenerate config' do
      allow(File).to receive(:open).and_call_original
      allow(File).to(receive(:open).with('/etc/postfix/main.cf') do
        double = double(File)
        allow(double).to receive(:read).and_return 'some content'
        double
      end)

      expect(chef_run).to create_if_missing_file('copy default main.cf to dropfolder').with(
        path: '/etc/postfix/main.cf.d/010-main.cf',
        content: 'some content'
      )
      expect(copy_main_cf).to notify('execute[generate config]').to(:run).delayed
    end

    it 'will install chef header and regenerate config' do
      expect(chef_run).to create_template('apply chef header config').with(
        path: '/etc/postfix/main.cf.d/000-chef_header.cf',
        source: 'postfix/main.cf.d/000-chef_header.cf',
        owner: 'root',
        group: 'root'
      )
      expect(install_chef_header).to notify('execute[generate config]').to(:run).delayed
    end

    it 'will install chef local.nf and regenerate config' do
      expect(chef_run).to create_template('apply local config').with(
        path: '/etc/postfix/main.cf.d/100-local.cf',
        source: 'postfix/main.cf.d/100-local.cf.erb',
        owner: 'root',
        group: 'root'
      )
      expect(install_local_cf).to notify('execute[generate config]').to(:run).delayed
    end

    it 'will install local destination table and restart postfix' do
      expect(chef_run).to create_cookbook_file('copy destination table for local config').with(
        path: '/etc/postfix/mydestinations',
        source: 'postfix/mydestinations'
      )
      expect(install_local_destinations_table).to notify('service[postfix]').to(:restart).delayed
    end

    it 'starts and enables postfix service' do
      expect(chef_run).to start_service('postfix')
      expect(chef_run).to enable_service('postfix')
    end
  end
end


test/smoke/default/default_test.rb
# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

control 'localmail-1.0' do
  title 'Install mailserver for localmail'
  desc 'Install a mailserver, make sure mails can be sent but are only delivered locally'

  sleep(5) # adding sleep to give postfix some time to start

  describe service('postfix') do
    it { should be_installed }
    it { should be_enabled }
    it { should be_running }
  end

  describe command('printf "Subject: This is a test\nMy message" | sudo sendmail test@test.de') do
    its('exit_status') { should eq 0 }
  end

  describe file('/var/mail/root') do
    it { should exist }
    its('content.strip') { should match(/Subject: This is a test/) }
  end
end


resources/mailhog.rb
# frozen_string_literal: true

unified_mode true

property :webui_port,
         Integer,
         default: 8025,
         description: 'The local port from where the web ui will be available'
property :sendmail_install_path,
         String,
         default: '/usr/sbin/sendmail',
         description: 'The path to where the mhsendmail binary will be downloaded, should be an executable path'

action :install do
  codenamephp_docker_service 'Install docker'

  docker_image 'pull mailhog image' do
    action :pull_if_missing
    repo 'mailhog/mailhog'
  end

  docker_container 'run mailhog' do
    container_name 'mailhog'
    repo 'mailhog/mailhog'
    port ['1025:1025', "#{new_resource.webui_port}:8025"]
    restart_policy 'always'
  end

  package 'remove sendmail package' do
    package_name 'sendmail'
    action :purge
  end

  template 'install sendmail for mailhog' do
    source 'mailhog/sendmail.erb'
    cookbook 'codenamephp_localmail'
    path new_resource.sendmail_install_path
    owner 'root'
    group 'root'
    mode '0755'
  end

  group 'add www-data user to docker group' do
    action :manage
    append true
    group_name 'docker'
    members 'www-data'
  end
end

action :uninstall do
  codenamephp_docker_service 'Install docker'

  docker_container 'remove mailhog container' do
    container_name 'mailhog'
    action %i(stop delete)
  end

  docker_image 'remove mailhog image' do
    action :remove
    repo 'mailhog/mailhog'
  end

  file 'remove sendmail' do
    path new_resource.sendmail_install_path

    action :delete
  end
end


recipes/postfix.rb
#
# Cookbook:: codenamephp_localmail
# Recipe:: postfix
#
# Copyright:: 2018, The Authors, All Rights Reserved.
#

include_recipe 'apt'

package 'postfix' do
  response_file 'postfix/preseed.erb'
end
package 'postfix-pcre'

package 'make'

cookbook_file 'install makefile for managing configurations' do
  path '/etc/postfix/Makefile'
  source 'postfix/Makefile'
  owner 'root'
  group 'root'
  mode '0755'
  notifies :run, 'execute[generate config]', :delayed
end

execute 'generate config' do
  command 'make'
  cwd '/etc/postfix'
  user 'root'
  action :nothing # only when notified
  notifies :restart, 'service[postfix]', :delayed
end

directory 'create main.cf.d dropfolder' do
  path '/etc/postfix/main.cf.d'
  mode 0o755
end

file 'copy default main.cf to dropfolder' do
  path '/etc/postfix/main.cf.d/010-main.cf'
  content(lazy { ::File.open('/etc/postfix/main.cf').read })
  action :create_if_missing
  notifies :run, 'execute[generate config]', :delayed
end

template 'apply chef header config' do
  path '/etc/postfix/main.cf.d/000-chef_header.cf'
  source 'postfix/main.cf.d/000-chef_header.cf'
  owner 'root'
  group 'root'
  notifies :run, 'execute[generate config]', :delayed
end

template 'apply local config' do
  path '/etc/postfix/main.cf.d/100-local.cf'
  source 'postfix/main.cf.d/100-local.cf.erb'
  owner 'root'
  group 'root'
  notifies :run, 'execute[generate config]', :delayed
end

cookbook_file 'copy destination table for local config' do
  path '/etc/postfix/mydestinations'
  source 'postfix/mydestinations'
  notifies :restart, 'service[postfix]', :delayed
end

service 'postfix' do
  action %i[enable start]
end
